go/pools/smartconnpool/pool.go

```diff
@@ -241,8 +265,8 @@ func (pool *ConnPool[C]) open() {
 // Open starts the background workers that manage the pool and gets it ready
 // to start serving out connections.
 func (pool *ConnPool[C]) Open(connect Connector[C], refresh RefreshCheck) *ConnPool[C] {
-	if pool.close != nil {
-		// already open
+	if !pool.state.CompareAndSwap(UNINITIALIZED, OPENED) {
+		// already open or closed
 		return pool
 	}
```

Comment:
`Open` probably should be changed to return an error when it's called on a pool object that has been closed previously. Unfortunately, that'd require changing the signature and call sites and would bloat up this PRs size.

Maybe that can be done as a followup step.